### PR TITLE
Add agent cache support for streaming endpoint

### PR DESCRIPTION
**Pull request by kyhavlov**
This PR adds a cachetype that supports using the streaming HealthServices gRPC endpoint for fetching data. There's still a couple remaining items left to do here (plus some cleanup of minor things):

- [x] Fill out the response protobuf structs with the missing fields
- [x] Add tests for the cachetype logic
- [x] Add ACL filtering to the RPC endpoint to match the existing endpoint's logic.

### CODE REVIEW

File: agent/consul/stream/health_endpoint.proto [hunk 1]
```diff
@@ -20,6 +20,7 @@ message Event {
     string key = 2;
     uint64 index = 3;
     ServiceHealthUpdate serviceHealth = 4;
+    bool endOfSnapshot = 5;
```
Comment (banks) [inline agent/consul/stream/health_endpoint.proto]: As noted above, maybe we should just make this a "snapshot" bool and level trigger rather than edge trigger?

I realise that it becomes more bits on the wire if every snapshot event or every non-snapshot even needs it to be set - few more bytes per message but that seems OK to me.

Alternatively (don't feel strongly) when we make Event type polymorphic, "EndOfSnapshot" should just become one of the types - I thought this was intentionally piggybacking on an existing event message but it's not in current usage then it saves having this single-purpose flag here.

File: agent/consul/streaming_endpoint.go [hunk 2]
```diff
@@ -29,37 +29,31 @@ func (h *ConsulGRPCAdapter) Subscribe(req *stream.SubscribeRequest, server strea
 	snapshotCh := make(chan stream.Event, 32)
 	go snapshotFunc(server.Context(), snapshotCh, req.Key)
 
+	h.srv.logger.Printf("[INFO] consul: new subscriber to %s/%s", req.Topic, req.Key)
```
Comment (banks) [inline agent/consul/streaming_endpoint.go]: Debug?

File: agent/consul/stream/health_endpoint.proto (hunk 1 above)
Comment (kyhavlov) [reply agent/consul/stream/health_endpoint.proto]: The only issue with this is that the client needs to know when the snapshot has ended before receiving any other messages (so that it can send the state to the watcher). With a snapshot = bool field there's no way to know when the snapshot is done being sent and it relies on getting another message that's an actual update before it can return something.

File: agent/consul/server.go [hunk 3]
```diff
@@ -756,6 +756,15 @@ func (s *Server) setupGRPC() error {
...
 	s.GRPCListener = lis
+
+	// Set up a local gRPC connection.
+	conn, err := grpc.Dial(lis.Addr().String(), grpc.WithInsecure(), grpc.WithDialer(dialGRPC), grpc.WithDisableRetry())
+	if err != nil {
+		return err
+	}
+
+	s.grpcConn = conn
+
```
[the remark this answers is not in the document]
Comment (kyhavlov) [reply agent/consul/server.go]: This is just for the delegate on the agent to get a grpc client with a connection to the local grpc server - there wasn't a way to create a client from the in-memory server.

**Comment by kyhavlov** [comment]
@banks Everything is pushed up now except for the cachetype test and the ACL filtering in the endpoint. I addressed the comments you left about simplifying the cache logic and pulled out the reusable parts for future cachetypes into separate code. I decided that made sense to do before the test since it would've been a lot of rewriting anyway. All the struct fields are added now too, so the responses are identical to the existing API.

File: agent/cache-types/streaming.go [hunk 4]
```diff
@@ -2,11 +2,174 @@ package cachetype
...
+		handler:  handler,
+		resultCh: make(chan interface{}, 0),
+	}
+
+	return s
+}
+
+// Close stops the subscriber by cancelling the context.
+func (s *Subscriber) Close() error {
+	s.cancelFunc()
```
[the remark this answers is not in the document]
Comment (banks) [reply agent/cache-types/streaming.go]: Oh also, it's probably unnecessary but consider nilling out the Handler and client here too to ensure that that state all gets garbage collected once it's dropped from the Cache.

File: agent/cache-types/streaming_health_services.go [hunk 5]
```diff
@@ -22,170 +20,65 @@ const (
...
-
-	// If something went wrong setting up the stream, return the error.
-	if err != nil {
-		// Do a blocking send since there should always be a Fetch goroutine watching
-		// on the startup of a subscriber.
-		s.resultCh <- err
-		return
+// HandleEvent updates the handler's state based on register/deregister events.
+func (h *healthServicesHandler) HandleEvent(event *stream.Event) {
+	serviceHealth := event.GetServiceHealth()
```
Comment (banks) [inline agent/cache-types/streaming_health_services.go]: Does this panic or return nil if the event was wrong type?

File: agent/cache-types/streaming_health_services.go [hunk 6]
```diff
@@ -22,170 +20,65 @@ const (
...
+	node := serviceHealth.ServiceNode
+	id := fmt.Sprintf("%s/%s", node.Node.Node, node.Service.ID)
+
+	switch serviceHealth.Op {
+	case stream.CatalogOp_Register:
+		checkServiceNode := stream.FromCheckServiceNode(serviceHealth.ServiceNode)
+		h.state[id] = checkServiceNode
```
Comment (banks) [inline agent/cache-types/streaming_health_services.go]: This is being called from the Subscriber goroutine but State below is being accessed by the Fetch goroutine. I think we need to synchronize map accesses here.

File: agent/cache-types/streaming.go [hunk 7]
```diff
@@ -2,11 +2,174 @@ package cachetype
...
 	"github.com/hashicorp/consul/agent/consul/stream"
 	"google.golang.org/grpc"
 )
 
-type GRPCClient interface {
+type StreamingClient interface {
 	Subscribe(ctx context.Context, in *stream.SubscribeRequest, opts ...grpc.CallOption) (stream.Consul_SubscribeClient, error)
 }
+
+type EventHandler interface {
```
Comment (banks) [inline agent/cache-types/streaming.go]: Maybe document that HandleEvent and State are called from different goroutines and so should syncronise access to any internal state they need.

File: agent/consul/health_endpoint.proto [hunk 8]
```diff
@@ -0,0 +1,15 @@
+syntax = "proto3";
+
+package consul;
+
+service Health {
```
Comment (banks) [inline agent/consul/health_endpoint.proto]: Is this just a testing stub? Maybe we could name it that way then? the real endpoint is just the Consul service now right?

File: agent/consul/health_endpoint.go [hunk 9]
```diff
@@ -16,6 +17,19 @@ type Health struct {
...
+}
+
+func (h *HealthGRPCAdapter) Stream(in *TestRequest, stream Health_StreamServer) error {
+	for {
+		if err := stream.Send(&TestReply{Data: int32(time.Now().Second())}); err != nil {
+			return err
+		}
+		time.Sleep(1 * time.Second)
+	}
+}
```
Comment (banks) [inline agent/consul/health_endpoint.go]: This seems to be test-only code but it's not named that way which is confusing. Why do we need it? All streaming should go through the Consul.Subscribe endpoint right?

Comment (banks) [reply agent/consul/health_endpoint.go]: Oh maybe this is old PoC code - can we kill it now?

File: agent/consul/server.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (banks) [reply agent/consul/server.go]: Ohhhhh. I missed that. Probably me being slow esp. with the `local gRPC connection` comment but if you feel like making it even more clear that you mean a connection to the built-in server in the same process then cool 😄 .

File: agent/consul/state/event.go [hunk 10]
```diff
@@ -49,6 +49,7 @@ func (e *EventPublisher) Commit() {
 				continue
 			}
 
+			// todo(kyhavlov): handle this better
```
Comment (banks) [inline agent/consul/state/event.go]: We should add this to the project where we are tracking issues. I don't think this is OK for release - missing events is not great!

File: agent/consul/stream/proto_types.go [hunk 11]
```diff
@@ -0,0 +1,76 @@
+package stream
+
+import (
+	"encoding/json"
+
+	proto "github.com/golang/protobuf/proto"
+)
+
+type UntypedMap map[string]interface{}
```
Comment (banks) [inline agent/consul/stream/proto_types.go]: What is this? It's not generated? I guess we maybe use this to represent opaque maps somehow but why is it not protobuf generated and why is it JSON encoding? Maybe some doc comments would clarify!

File: agent/consul/stream/structs.proto [hunk 12]
```diff
@@ -0,0 +1,377 @@
...
+  string Address = 3;
+  string Datacenter = 4;
+  map<string, string> TaggedAddresses = 5;
+  map<string, string> Meta = 6;
+
+  RaftIndex RaftIndex = 7
+      [(gogoproto.embed) = true, (gogoproto.nullable) = false];
+}
+
+// ServiceDefinition is used to JSON decode the Service definitions. For
```
Comment (banks) [inline agent/consul/stream/structs.proto]: Is this copy pasta from the original struct? Maybe if this file was generated we should add a comment to the top to make it clear?

File: agent/consul/stream/subscribe.proto [hunk 13]
```diff
@@ -2,12 +2,15 @@ syntax = "proto3";
...
 
+import "structs.proto";
+
 service Consul {
     rpc Subscribe(SubscribeRequest) returns (stream Event) {}
 } 
 
 enum Topic {
     ServiceHealth = 0;
+    EndOfSnapshot = 1;
```
Comment (banks) [inline agent/consul/stream/subscribe.proto]: I really don't think this should be a topic. The semantics of a topic are "a stream of events all about one thing" when we subscribe to `ServiceHealth` topic I think it's odd to use the topic field to mark "end of snapshot" events within the topic.

Imagine if we extend this to allow subscribing to multiple topics in one call, now they each need to send "end of snapshot" messages but we need to know which topic each one is for so we can't use that.

I think we should add an `EndOfSnapshot` message type which is otherwise an empty struct and make it one of the `payload` types below and then type check in the subscriber loop.

File: agent/consul/streaming_endpoint.go [hunk 14]
```diff
@@ -30,36 +30,28 @@ func (h *ConsulGRPCAdapter) Subscribe(req *stream.SubscribeRequest, server strea
...
+			return err
 		}
 	}
 
+	// Send a marker that this is the end of the snapshot.
+	endSnapshotEvent := stream.Event{Topic: stream.Topic_EndOfSnapshot}
+	if err := server.Send(&endSnapshotEvent); err != nil {
+		return err
+	}
+	h.srv.logger.Printf("[INFO] consul: finished sending snapshot to new subscriber")
```
Comment (banks) [inline agent/consul/streaming_endpoint.go]: Should probably be DEBUG.

File: agent/consul/health_endpoint.proto (hunk 8 above)
Comment (kyhavlov) [reply agent/consul/health_endpoint.proto]: I thought this was removed, it looks like it came back during rebasing somehow

File: agent/cache-types/streaming_health_services.go (hunk 6 above)
Comment (kyhavlov) [reply agent/cache-types/streaming_health_services.go]: both are only called in Subscriber.run(), unless you're referring to something else?

File: agent/cache-types/streaming_health_services.go (hunk 5 above)
Comment (kyhavlov) [reply agent/cache-types/streaming_health_services.go]: I changed the subscriber to panic if it somehow receives an event of the wrong type

Comment (banks) [reply agent/cache-types/streaming_health_services.go]: How does subscriber know that it's the wrong type? this is the only place that knows which types the health topic should be sending.

I think panic is also dramatic too - it makes any bug with backward compatibility later into a "crash all agents" instead of just" some blocking queries broke. I suggest we error (not just ignore as that leads to silent bad state) the stream but not panic the whole agent.

File: agent/cache-types/streaming_health_services.go (hunk 6 above)
Comment (banks) [reply agent/cache-types/streaming_health_services.go]: 🤔 Think I was confused with the `State` variable in the Fetcher.

File: agent/consul/acl.go [hunk 15]
```diff
@@ -1087,6 +1094,37 @@ func (f *aclFilter) allowSession(node string) bool {
...
+		}
+		val := f.allowCheckServiceNode(node, service)
+		f.logger.Printf("checking allow on %s/%s (%v)", node, service, val)
+		return val
+	default:
+		f.logger.Printf("could not filter for unrecognized event topic %s", event.Topic)
+		return true
+	}
+}
+
```
Comment (banks) [inline agent/consul/acl.go]: Hmm preferably type switching on specific event types should be as localized as possible. Otherwise implementing a new type requires code in lots of different places.

Could we maybe make this a method on `Event` so that this code is generic and all the type-specifics are contained in the Event file alongside the other things you'd have to change to add a new event type?

File: agent/consul/stream/subscribe.proto [hunk 16]
```diff
@@ -13,13 +15,17 @@ enum Topic {
 message SubscribeRequest {
     Topic topic = 1;
     string key = 2;
+    string token = 3;
```
Comment (banks) [inline agent/consul/stream/subscribe.proto]: Do we need DC in here? I've not thought about cross-DC forwarding at all yet but we probably should. I suggest making a note of it on the project board in Asana and we can tackle that separately.

File: agent/consul/acl.go (hunk 15 above)
Comment (kyhavlov) [reply agent/consul/acl.go]: Event is all generated so you'd have to go to two places anyway  (adding the proto declarations + acl logic file) - I put this here to reuse the `aclFilter` logic for existing things. I can move the aclFilter stuff to its own package and make it public to call it from the stream package if you think that's cleaner though.

**Comment by kyhavlov** [comment]
They pass, but only the cache gets tested - we don't have any for a blocking query given a specific index, so I'll add one for that.

File: agent/consul/acl.go (hunk 15 above)
Comment (banks) [reply agent/consul/acl.go]: Ah yeah forgot it's generated.

Do you remember my idea about how we should have each event type be able to encode it's own set of ACL requirements so that we can un-ambiguously enforce them on client side later without compatibility issues etc? If we think along those lines then I think it suggest the same cleaner separation as I was getting at here:

 1. If possible, use some annotation in the protobuf and an extension or something to encode the ACL requirements by type (with `topic` and `key` as optional inputs?). If not possible, define a go file in the same place as the proto file that has a method that encodes this (so all are encoded in one place and any other type specialisation can go there too)
 2. have this method just call that method, retreive the set of things to enforce and validate that.

What do you think?

Comment (kyhavlov) [reply agent/consul/acl.go]: Putting the ACL requirements in an explicit field makes sense so it's unambiguous to clients later - I just pushed up a change to do that.

File: agent/consul/stream/subscribe.proto [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+
+enum ACLResource {
+    ServiceACL = 0;
+    NodeACL = 1;
+}
+
+message ACLRule {
+    ACLResource Resource = 1;
+    string Segment = 2; 
+}
```
Comment (banks) [inline agent/consul/stream/subscribe.proto]: @mkeeler do you have an opinion on this representation and/or idea of embedding ACL rules into events?

The tl/dr is:
 - We want to define in one central place which types of ACL rules need to be checked for each event type.
 - Since events are generated by protobuf we have to do that somewhere in code (acl.go in this file needs to be updated along side event definitions but still all in one place).
 - Encoding them on the wire has a really nice future benefit - we can un-ambiguously enforce ACL at the client without needing to worry about compatibility between client and server versions etc. letting new types slip through or different rules be applied. This means that in the future clients can optimize and de-duplicate cache entries from multiple differnt clients while still ensureing they filter results according to the ACL rules the server would have applied.

This seems like a reasonable minimal represenation for those and it's internal only and can be evolved as we build more types and extend ACLs but interested if you see issues with this idea or encoding?

File: agent/consul/streaming_endpoint.go [hunk 18]
```diff
@@ -29,37 +29,49 @@ func (h *ConsulGRPCAdapter) Subscribe(req *stream.SubscribeRequest, server strea
...
-			if !ok {
-				h.srv.logger.Printf("[INFO] consul: finished sending snapshot")
-				snapshotDone = true
-				break
-			}
-			if err := server.Send(&event); err != nil {
-				return err
-			}
+	for event := range snapshotCh {
+		event.SetACLRules()
```
Comment (banks) [inline agent/consul/streaming_endpoint.go]: Can we just have `allowEvent` call this? Then we don't have to remember to prime it everywhere we might filter an event to ensure it's set properly?

It would be kinda nice if there was an easy way to have protoc generate a hard coded method into each event type that just returns the correct set somehow rather than need an extra method but I think that's probably a complication we don't need to work on just now.

Comment (banks) [reply agent/consul/streaming_endpoint.go]: Realised that doesn't work because then calling `filter.applyEvent` on a client would override the required ACLs with the ones the client code knew about not the ones the server sent on the wire.

🤔 There is something a little bit gross about being able to accidentally do this. I wonder if there is a way around this or at least make it a little less footgunny?

File: agent/consul/acl.go [hunk 19]
```diff
@@ -1087,6 +1094,32 @@ func (f *aclFilter) allowSession(node string) bool {
...
 }
 
+// allowEvent is used to determine if a stream.Event is accessible for an ACL.
+func (f *aclFilter) allowEvent(event stream.Event) bool {
+	// Fast path if ACLs are not enabled
+	if f.authorizer == nil {
+		return true
+	}
+
+	allow := true
```
Comment (banks) [inline agent/consul/acl.go]: Hmm I wonder if we could default to deny and then have an explicit `no ACLs required` sentinel value to encode in the protobuf so that we can reliably determine the difference between "no ACLs required" and "forgot to call the method that sets up the right ACLs in the event"?

File: agent/consul/streaming_endpoint.go (hunk 18 above)
Comment (kyhavlov) [reply agent/consul/streaming_endpoint.go]: Yeah it does seem weird that the required acl permissions could be updated separately, but we shouldn't need to modify the event at all outside of this endpoint. I don't think having to worry about calling SetACLRules will be an issue either - we only ever have to call that here in Subscribe on the server.

File: agent/consul/stream/subscribe.proto (hunk 17 above)
Comment (mkeeler) [reply agent/consul/stream/subscribe.proto]: If I understand this PR correctly, we will be enforcing ACLs server side so that we don't send events to clients that shouldn't see them but we want to also send the required rules to the client so that it can deliver the final results to all those http clients that are waiting for them and that have permissions to have the data.

If that is correct then I believe that this change is a good one however I don't believe it will be a silver bullet for future compatibility. If we add new ACL resources in the future the clients will still need to be updated to be made aware of those things. That is unless we rework the github.com/hashicorp/consul/acl package to handle resources/segments more generically. Then again the clients HTTP API wouldn't be able to handle those other types anyways so that probably doesn't matter. What it does buy us is the ability to change which available types of resources are required for enforcement. I could see this being especially important for config entry streaming eventually.

A couple other more general questions are:

1. Have we put any thought into how this would interact with Sentinel. Because the tokens/policies will get resolved both on the client and server the sentinel policy would execute on both. Streaming shouldn't affect it but it might be worth verifying.

2. Maybe I missed it in this PR (or potentially another one), but how do we handle multiple subscriptions using different tokens. Do we open up a new stream for the different token? Once streaming has started how does a client request subscriptions to additional topics?

Having not thought about the details nearly as much as either of you I would have expected something like:

1. open GRPC stream with some initial topic(s) and token(s)
2. server sends snapshot of the event data for the topic(s) filtering the events by the token(s)
3. server now streams events to the client
4. client requests subscription for new topic(s) with token(s)
5. server sends snapshot of the new data for the topic(s) filtering the events by the token(s)
6. Repeat 3-5 for a while
7. client unsubscribes from topic(s) and/or removes/updates token(s)
8. Back to 3-5 again
9. Eventually the stream gets closed

Basically with one stream there is a back and forth between the server and client to (un)subscribe to topic(s) and manage the ACL tokens used by the stream as many could be and probably will be involved.

Comment (banks) [reply agent/consul/stream/subscribe.proto]: Great points Matt thanks for the sanity check.

> Have we put any thought into how this would interact with Sentinel. 

Nope. For now the ACL embedding is a future-proofing mechanism. I was actually only intending to build the mechanism to define those so this was possible later but I guess it doesn't hurt to actually transfer them already too. Sentinel would need to be executed on client too otherwise we'd need to special case ACLs that have sentinel attached and not allow that client side. This is less a comitment to doing things client side and more just an idea that makes the code here a little cleaner (all ACL needs in one place near type definitions) and also allows for future _possibility_ of more efficient clients with shared cache entries.

> Maybe I missed it in this PR (or potentially another one), but how do we handle multiple subscriptions using different tokens. Do we open up a new stream for the different token?

Yes. Agent cache is partitioned by Token implicitly so every client with a different token gets a separate cache entry for a given request currently. Obviously that might be wasteful when they all are the same and there are many clients/tokens on a single agent but that's not the common case and that is the simplest model for not leaking stuff. This addition was a step towards potentially being able to relax that in some cases later if we think that's worth the extra complexity.

> Once streaming has started how does a client request subscriptions to additional topics?

It doesn't. The current model is super simple - each `Subscribe` RPC is for a single topic. If a client needs multiple topics each one is a separate cache entry and a separate backend `Subscribe` call. In some cases that may mean the same event is sent redundantly on multiple subscriptions to the same client but the simplicity makes this a good trade - especial give the fact we are already several orders of magnitude better than before streaming.

We thought about more complex client behaviour where the client maintains just a single subscription and dynamically subscribes and unsubscribes to topics but this hugely complicates the client code and also snapshotting and playback semantics per topic on a shared connect etc. Unless the performance is a real-world issue I doubt we'd want to go there.

File: agent/health_endpoint.go [hunk 20]
```diff
@@ -186,7 +186,7 @@ func (s *HTTPServer) healthServiceNodes(resp http.ResponseWriter, req *http.Requ
 	var out structs.IndexedCheckServiceNodes
 	defer setMeta(resp, &out.QueryMeta)
 
-	if args.QueryOptions.UseCache || args.MinQueryIndex > 0 {
+	if (args.QueryOptions.UseCache || args.MinQueryIndex > 0) && !connect {
```
Comment (banks) [inline agent/health_endpoint.go]: Why only non-connect?

File: agent/consul/stream/subscribe.proto (hunk 17 above)
Comment (mikemorris) [reply agent/consul/stream/subscribe.proto]: This is very much a naïve question looking at this without context, but are we signing the contents of the Protobuf messages in some way to protect against something like a spoofed message embedding `required_acls: []`?

Comment (banks) [reply agent/consul/stream/subscribe.proto]: Great question @mikemorris. We aren't because we rely on TLS for any secure communication.

This is a response from a server that we should be talking to over TLS and should have validated its identity on connection. If we don't to either of those then games already over for any MITM attack that wants to send us arbitrary stuff.

File: agent/health_endpoint.go (hunk 20 above)
Comment (kyhavlov) [reply agent/health_endpoint.go]: When `connect = true` we get effectively a different RPC endpoint (different state store logic/table lookup) so I think it'll require a different topic. The current endpoint shares filtering logic since the return types are the same but swaps the state store lookup function based on whether or not connect is enabled.

**Comment by kyhavlov** [comment]
@banks I was planning to do cross-DC requests in another PR for ease of reviewing since this one's gotten pretty long. I haven't looked into it much yet so if it turns out to be non-trivial I didn't want it to hold up merging this PR.

File: agent/health_endpoint.go (hunk 20 above)
Comment (banks) [reply agent/health_endpoint.go]: 🤔 It would be _really_ nice to support connect at launch since that is one of the biggest use-cases for discovery in our service mesh model.

I think you're right about separate topic though so it seems OK to do that in a separate PR but maybe add a TODO to the project board?

**Comment by banks** [comment]
Cool. Lets merge this in to the feature branch and just make sure cross-DC and Connect are on the TODO list for launch?
